src/main/java/edu/unc/lib/boxc/migration/cdm/services/PermissionsService.java

```diff
@@ -295,20 +297,47 @@ private List<List<String>> updateCsvRecords(PermissionMappingOptions options) {
         String everyoneField = getAssignedRoleValue(options.isStaffOnly(), options.getEveryone());
         String authenticatedField = getAssignedRoleValue(options.isStaffOnly(), options.getAuthenticated());
 
-        // update existing entry
-        for (CSVRecord record : previousRecords) {
-            cdmIds.add(record.get(0));
-            if (record.get(0).equals(options.getCdmId())) {
-                updatedRecords.add(Arrays.asList(record.get(0), record.get(1), everyoneField, authenticatedField));
-            } else {
-                updatedRecords.add(Arrays.asList(record.get(0), record.get(1), record.get(2), record.get(3)));
+        // add or update individual entry
+        if (options.getCdmId() != null) {
+            // update existing entry
+            for (CSVRecord record : previousRecords) {
+                cdmIds.add(record.get(0));
+                if (record.get(0).equals(options.getCdmId())) {
+                    updatedRecords.add(Arrays.asList(record.get(0), record.get(1), everyoneField, authenticatedField));
+                } else {
+                    updatedRecords.add(Arrays.asList(record.get(0), record.get(1), record.get(2), record.get(3)));
+                }
+            }
+            // add new entry
+            if (!cdmIds.contains(options.getCdmId())) {
+                String objectType = getObjectType(options.getCdmId());
+                updatedRecords.add(Arrays.asList(options.getCdmId(), objectType, everyoneField, authenticatedField));
             }
         }
 
-        // add new entry
-        if (!cdmIds.contains(options.getCdmId())) {
-            String objectType = getObjectType(options.getCdmId());
-            updatedRecords.add(Arrays.asList(options.getCdmId(), objectType, everyoneField, authenticatedField));
+        // add or update with-works and with-files entries
+        if (options.isWithWorks() || options.isWithFiles()) {
```

Comment:
I haven't confirmed it, but i think if cdmId is set and either one of these options is set, you'll end up with a CSV document with everything twice since the cdmId workflow fills everything in and it all gets added again below. 

I'm wondering if it might make sense to integrate the two workflows by compiling a list of all the ids that need to be updated/added (cdmId + the outcome from addWorkFileRecords), then use the workflow already present for cdmId (loop through the csv, if the id is in the update list then change it otherwise re-add it, and then add all the new stuff at the end). Might be shorter? Alternatively, you could make this block into an `else if` of the cdmId block, in which case the two workflows couldn't happen at the same time. The implementation you currently have seems like it should work well enough otherwise.